lib/modules/manager/pep621/processors/uv.ts

```diff
@@ -32,6 +33,13 @@ export class UvProcessor implements PyProjectProcessor {
       ),
     );
 
+    const registryUrls = extractRegistryUrls(uv);
+    if (registryUrls.length) {
+      for (const dep of deps) {
+        dep.registryUrls = [...registryUrls];
+      }
```

Comment:
Normally if registry URLs apply to the entire package file then we want them set once at the packageFile level instead duplicated into every dep.

Also, are these likely to _all_ be pypi datasource now and forever? Or is it possible that there could be a git dependency mixed in there at one point (which would then make the registry URL invalid for that dep)

Reply:
> Normally if registry URLs apply to the entire package file then we want them set once at the packageFile level instead duplicated into every dep.

There is a plan to be able to map specific dependencies to specific sources in https://github.com/astral-sh/uv/issues/171, similarly to what PDM does I believe. According to [this comment](https://github.com/astral-sh/uv/pull/6389#issuecomment-2327284874) it should be worked on soon enough. It doesn't mean that we can't implement what you're suggesting, which makes sense, but maybe since this is meant to evolve, setting registries at the dependency level is more conservative?

> Also, are these likely to _all_ be pypi datasource now and forever? Or is it possible that there could be a git dependency mixed in there at one point (which would then make the registry URL invalid for that dep)

That's a really good point, even today it's already possible to set other sources, like [git](https://docs.astral.sh/uv/concepts/dependencies/#git), [URL](https://docs.astral.sh/uv/concepts/dependencies/#url), [path](https://docs.astral.sh/uv/concepts/dependencies/#path), or [workspace](https://docs.astral.sh/uv/concepts/dependencies/#workspace-member) dependencies. We should definitely skip those for now similarly to [what we do in `cargo`](https://docs.astral.sh/uv/concepts/dependencies/#workspace-member), by checking if a source is defined for a dependency (and later on we could support updates for git dependencies for instance).

I can make a separate PR to handle skipping those dependencies, since I believe this is not directly related to what this PR implements, as today we already consider dependencies with different sources as public PyPI ones anyway.

Reply:
Created https://github.com/renovatebot/renovate/pull/31270 for the 2nd point, thanks for pointing that out, I should have been implemented from the start.